File: backend/agent_03_content_auditor.py
import os
import redis
import json
from supabase import create_client, Client
import google.generativeai as genai

# Environment variables
SUPABASE_URL = os.environ.get("SUPABASE_URL", "http://localhost:8000")
SUPABASE_KEY = os.environ.get("SUPABASE_KEY", "dummy_key")
REDIS_URL = os.environ.get("REDIS_URL", "redis://localhost:6379")
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY", "")

# Initialize clients
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
redis_client = redis.Redis.from_url(REDIS_URL)

if GEMINI_API_KEY:
    genai.configure(api_key=GEMINI_API_KEY)
# llm_model = genai.GenerativeModel('gemini-2.5-flash')

def fetch_gsc_data(url):
    """
    Mock Google Search Console data fetching.
    """
    print(f"Agent 03: Fetching GSC data for {url}...")
    # Return simulated CTR, position, clicks, etc.
    return {
        "avg_position": 15.2,
        "clicks": 142,
        "ctr": 1.2,
        "impressions": 11800,
        "content_freshness": "low",
        "word_count_diff": -300
    }

def calculate_performance_score(gsc_data, seo_score):
    """
    Calculate an overall performance score based on multiple factors.
    """
    print("Agent 03: Calculating performance score...")
    # Base logic mapping real metrics to a 0-100 score. 
    # Example logic:
    score = 100
    if gsc_data["ctr"] < 3.0:
        score -= 15
    if gsc_data["avg_position"] > 10:
        score -= 20
    if gsc_data["content_freshness"] == "low":
        score -= 15
    if gsc_data["word_count_diff"] < 0:
        score -= 10
        
    final_score = (score + seo_score) / 2
    return int(final_score)

def create_rewrite_brief(post_id, post_data, metrics, score):
    """
    Use LLM to generate a rewriting guide/brief.
    """
    print(f"Agent 03: Performance score {score} < 60. Creating rewrite brief for {post_id}...")
    
    # Store performance metrics
    supabase.table("post_performance").insert({
        "post_id": post_id,
        "avg_position": metrics["avg_position"],
        "clicks": metrics["clicks"],
        "ctr": metrics["ctr"],
        "impressions": metrics["impressions"],
        "score": score
    }).execute()
    
    # In a real app, you would create a new entry in a rewrite_briefs table or update content_briefs
    print(f"Agent 03: Sending rewrite task back to Agent 02...")
    
    event_data = {
        "event": "audit_complete",
        "post_id": post_id,
        "action": "rewrite",
        "reason": f"Low CTR ({metrics['ctr']}%) and poor rank."
    }
    redis_client.publish("content_events", json.dumps(event_data))

def process_audit(post_id):
    # Fetch post
    response = supabase.table("content").select("*").eq("id", post_id).execute()
    if not response.data:
        print(f"Agent 03: Post {post_id} not found!")
        return
        
    post = response.data[0]
    live_url = post['live_url']
    seo_score = post['seo_score']
    
    metrics = fetch_gsc_data(live_url)
    score = calculate_performance_score(metrics, seo_score)
    
    print(f"Agent 03: Calculated score {score} for post {post['title']}")
    
    if score < 60:
        create_rewrite_brief(post_id, post, metrics, score)
    else:
        print("Agent 03: Score is 60 or higher. No action needed.")

def listen_for_events():
    """
    Listen on Redis queue.
    """
    print("Agent 03 - Content Auditor listening for events...")
    pubsub = redis_client.pubsub()
    pubsub.subscribe("content_events")
    
    for message in pubsub.listen():
        if message['type'] == 'message':
            data = json.loads(message['data'])
            if data['event'] == 'post_published':
                post_id = data['post_id']
                process_audit(post_id)

if __name__ == "__main__":
    listen_for_events()

File: backend/agent_03_content_auditor.js
const { createClient } = require('@supabase/supabase-js');
const redis = require('redis');
const Groq = require('groq-sdk');
const { isValidRedisUrl } = require('./redis-helper');
require('dotenv').config({ path: '../frontend/.env' });

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_KEY;
const REDIS_URL = process.env.REDIS_URL || '';
const GROQ_API_KEY = process.env.GROQ_API_KEY;

const supabase = (SUPABASE_URL && SUPABASE_KEY && SUPABASE_URL.startsWith('http'))
    ? createClient(SUPABASE_URL, SUPABASE_KEY)
    : { from: () => ({ select: () => ({ eq: () => ({}) }), insert: () => ({ select: () => ({}) }), update: () => ({ eq: () => ({}) }) }) };
let redisClient;

const groq = GROQ_API_KEY ? new Groq({ apiKey: GROQ_API_KEY }) : null;

async function logActivity(agentName, eventType, message, metadata = {}) {
    try {
        if (!supabase) return;
        await supabase.from('agent_logs').insert({
            agent_name: agentName,
            event_type: eventType,
            message: message,
            metadata: metadata
        });
    } catch (err) {
        console.error('Logging failed:', err.message);
    }
}

async function fetchGSCData(url) {
    console.log(`Agent 03: Simulating dynamic GSC data for ${url}...`);
    return {
        avg_position: Number((Math.random() * 20 + 1).toFixed(1)), // 1.0 to 21.0
        clicks: Math.floor(Math.random() * 500),
        ctr: Number((Math.random() * 5).toFixed(1)), // 0.0 to 5.0
        impressions: Math.floor(Math.random() * 20000),
        content_freshness: Math.random() > 0.5 ? 'high' : 'low',
        word_count_diff: Math.floor(Math.random() * 600) - 300
    };
}

function calculatePerformanceScore(gscData, seoScore) {
    console.log('Agent 03: Calculating performance score...');
    let score = 100;
    if (gscData.ctr < 3.0) score -= 15;
    if (gscData.avg_position > 10) score -= 20;
    if (gscData.content_freshness === 'low') score -= 15;
    if (gscData.word_count_diff < 0) score -= 10;

    return Math.floor((score + seoScore) / 2);
}

async function createRewriteBrief(postId, postData, metrics, score) {
    console.log(`Agent 03: Performance score ${score} < 60. Alerting for rewrite for ${postId}...`);

    console.log(`Agent 03: Sending rewrite task back to Agent 02...`);

    const eventData = {
        event: 'audit_complete',
        post_id: postId,
        action: 'rewrite',
        reason: `Low CTR (${metrics.ctr}%) and poor rank.`
    };

    if (!isValidRedisUrl(REDIS_URL)) {
        console.warn('Agent 03: Redis disabled — skipping rewrite event emit.');
        return;
    }
    const publishClient = redis.createClient({ url: REDIS_URL });
    await publishClient.connect();
    await publishClient.publish('content_events', JSON.stringify(eventData));
    await publishClient.disconnect();
}

async function processAudit(postId) {
    try {
        const { data: responseData, error } = await supabase.from('content').select('*').eq('id', postId);

        if (error || !responseData || responseData.length === 0) {
            console.log(`Agent 03: Post ${postId} not found!`);
            return;
        }

        const post = responseData[0];
        const liveUrl = post.live_url;
        const seoScore = post.seo_score || 0;

        const metrics = await fetchGSCData(liveUrl);
        const score = calculatePerformanceScore(metrics, seoScore);

        console.log(`Agent 03: Calculated score ${score} for post ${post.title}`);

        // ALWAYS RECORD AUDIT DATA
        await supabase.from('post_performance').insert({
            post_id: postId,
            avg_position: metrics.avg_position,
            clicks: metrics.clicks,
            ctr: metrics.ctr,
            impressions: metrics.impressions,
            score: score
        });

        if (score < 60) {
            await createRewriteBrief(postId, post, metrics, score);
            await logActivity('Auditor (Agent 03)', 'ERROR', `SEO Alert! Low performance for: ${post.title}`, { score, postId });
        } else {
            console.log("Agent 03: Score is 60 or higher. No rewrite needed.");
            await logActivity('Auditor (Agent 03)', 'SUCCESS', `SEO Audit Complete for: ${post.title}`, { score, postId });
        }
    } catch (err) {
        console.error(`Error auditing post ${postId}:`, err);
        await logActivity('Auditor (Agent 03)', 'ERROR', `Audit failed for post ID ${postId}`, { error: err.message });
    }
}

async function listenForEvents() {
    console.log('Agent 03 - Content Auditor listening for events...');
    if (!isValidRedisUrl(REDIS_URL)) {
        console.warn('Agent 03: Redis disabled — REDIS_URL invalid. Auditor will be inactive.');
        return;
    }
    try {
        redisClient = redis.createClient({ url: REDIS_URL });
        await redisClient.connect();

        await redisClient.subscribe('content_events', (message) => {
            try {
                const data = JSON.parse(message);
                if (data.event === 'post_published') {
                    processAudit(data.post_id);
                }
            } catch (err) {
                console.error('Agent 03: Mismatch event data parsing', err);
            }
        });
    } catch (err) {
        console.error('Agent 03 listener failed:', err);
    }
}

async function runPeriodicAudit() {
    console.log('Agent 03: Starting periodic 4-hour SEO audit...');
    try {
        const { data: posts, error } = await supabase.from('content').select('id').eq('status', 'PUBLISHED');
        if (!error && posts) {
            for (const post of posts) {
                await processAudit(post.id);
            }
        }
    } catch (e) { console.error('Periodic audit fail', e); }
}

async function startAuditor() {
    listenForEvents();
    
    // Periodic audit every 4 hours
    while (true) {
        await logActivity('Auditor (Agent 03)', 'INFO', 'Starting system-wide SEO performance sweep');
        await runPeriodicAudit();
        console.log('Agent 03: Sweep complete. Sleeping for 4 hours...');
        await new Promise(res => setTimeout(res, 4 * 60 * 60 * 1000));
        // await new Promise(res => setTimeout(res, 4 * 60 * 60 * 1000));
    }
}

startAuditor();
